synapse/storage/databases/main/event_push_actions.py

```diff
@@ -12,6 +12,67 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+"""Responsible for storing and fetching push actions / notifications.
+
+There are two main uses for push actions:
+  1. Sending out push to a user's device; and
+  2. Tracking per-room per-user notification counts (used in sync requests).
+
+For the former we simply use the `event_push_actions` table, which contains all
+the calculated actions for a given user (which were calculated by the
+`BulkPushRuleEvaluator`).
+
+For the latter we could simply count the number of rows in `event_push_actions`
+table for a given room/user, but in practice this could be *very* heavyweight
+when there were a large number of notifications (due to e.g. the user never
+reading a room). Plus, keeping all push actions indefinitely uses a lot of disk
+space.
+
+To fix these issues, we add a new table `event_push_summary` that tracks
+per-user per-room count of all notifications that happened before a stream
+ordering S. Thus, to get the notification count for a user / room we can simply
+query a single row in `event_push_summary` and count the number of rows in
+`event_push_actions` with a stream ordering larger than S (and as long as S is
+"recent", the number of rows needing to be scanned will be small).
+
+The `event_push_summary` table is updated via a background job that periodically
+choses a new stream ordering S' (usually the latest stream ordering), counts all
+notifications in `event_push_actions` between the existing S and S', and adds
+them to the existing counts in `event_push_summary`.
+
+This allows us to delete old rows from `event_push_actions` once those rows have
+been counted and added to `event_push_summary` (we call this process
+"rotation").
+
+
+We need to handle when a user sends a read receipt to the room. Again this is
+done as a background process. For each receipt we clear the row in
+`event_push_summary` and count the number of notifications in
+`event_push_actions` that happened after the receipt but before S, and insert
+that count into `event_push_summary`.
```

Comment:
This paragraph assumes the read receipt is < `S`; I believe it can also happen after `S` in which case we completely clear the `event_push_summary` row and anything in `event_push_actions` which is <= R?

Reply:
Well, I think of `after the receipt but before S` as meaning the empty set if receipt is after `S`.  Will clarify

I have somewhat intentionally not said when we trigger removal of `event_push_actions`, as its a bit handwavey and doesn't really affect how this all works.

Reply:
Essentially this causes us to fallback to the "just use `event_push_actions` table", I think?

Reply:
Yup! Though by "fallback" here, that's just the same behaviour as normal (i.e. we always read from the event_push_actions to check for events that have happened since S)

Reply:
Yup! Though by "fallback" here, that's just the same behaviour as normal (i.e. we always read from the event_push_actions to check for events that have happened since S)

Reply:
Right. Makes sense!